pandas/core/arrays/masked.py

```diff
@@ -1098,25 +1108,62 @@ def _reduce(self, name: str, *, skipna: bool = True, **kwargs):
 
         # median, skew, kurt, sem
         op = getattr(nanops, f"nan{name}")
-        result = op(data, axis=0, skipna=skipna, mask=mask, **kwargs)
-
+        axis = kwargs.pop("axis", None)
+        result = op(data, axis=axis, skipna=skipna, mask=mask, **kwargs)
         if np.isnan(result):
-            return libmissing.NA
+            result = libmissing.NA
 
-        return result
+        return self._wrap_reduction_result(
+            name, result, skipna=skipna, axis=axis, **kwargs
+        )
+
+    def _reduce_and_wrap(self, name: str, *, skipna: bool = True, kwargs):
+        df = self.reshape(-1, 1)
+        res = df._reduce(name=name, skipna=skipna, axis=0, **kwargs)
+        return res
 
     def _wrap_reduction_result(self, name: str, result, skipna, **kwargs):
+        axis = kwargs["axis"]
         if isinstance(result, np.ndarray):
-            axis = kwargs["axis"]
             if skipna:
                 # we only retain mask for all-NA rows/columns
                 mask = self._mask.all(axis=axis)
             else:
                 mask = self._mask.any(axis=axis)
 
             return self._maybe_mask_result(result, mask)
+        elif result is libmissing.NA and self.ndim == 2:
+            result = self._wrap_na_result(name=name, axis=axis)
+            return result
         return result
 
+    def _wrap_na_result(self, *, name, axis):
```

Comment:
this feels like it got a lot more complicated than the first attempt at "keepdims".  does this just address more corner cases the first attempt missed?

Reply:
Yeah, this is not great, and I've been tinkered quite a lot with other implementations. I think/hope I will have a simpler solution tonight.

Reply:
Ok, I looked into this and unless I do a relatively major refactoring, it looks like I just move complexity around by changing this. So without a bigger rework of `arrays_algos`, I don't see any clearly better method. 

Suggestions/ideas that show otherwise welcome, of course.

Reply:
to add, the underlying issue is that the functions in `masked_reductions.py` only return scalar `NA` values. This means that we don't get the type information when doing reductions that return `NA`, but have to infer it and the inferring is complex because it depends on calling method, dtype of calling data and OS platform. 

The solution would be to return proper 2D results from `masked_reductions.py` functions, e.g. return a tuple instead of scalar, e.g.
1.  return (scalar_value, none), when returning scalar
2. return (value_array, mask_array) when returning a masked array

and the wrap the result in an array  in `BaseMaskedArray._wrap_reduction_result`, when it should be a masked array.

However, I 'm not sure if that's the direction we want to pursue because unless we want to support 2d masked arrays, the current solution could still be simpler than building out 2d reduction support.

Reply:
Longer term, that indeed sounds as a good alternative, even for the current 1D-reshaped-as-2D case. One potential disadvantage is that you still need to do some actual calculation on the data, unnecessarily, to get the `value_array` return value (so we can use that to determine the result dtype). Calculating the actual value is "unnecessary" if you know the result will be masked. Of course it avoids hardcoding result dtypes. But for the case of skipna=False with a large array, that might introduce a slowdown?

Reply:
We only "know" because of the hard-coding here, for example `Series([1, 2, pd.NA], dtype="int8").sum()` is `Int32` on some systems and `Int64` on others. So to avoid the hardcoding we will have to do a calculation, so it's a trade-off (or we could hard-code the dtype when we're 100 % sure we get a NA, but that'll be about the same complexity as now + the extra code for doing real 2d ops). I think we should only do the 2d ops if we want it independent of this issue here. 

IMO that's a possible future PR, if we choose to go in that direction.

Reply:
> We only "know" because of the hard-coding here

In the end, also numpy kind of "hard codes" this, but just in the implementation itself. And we get the difficulty of combining the algo of numpy with our own logic, where we also want to know this information. So I don't think it's necessarily "wrong" to hard code this on our side as well (i.e. essentially keep it as you are doing in this PR, also long term (at least as long as the arrays are only 1D)).

Sidenote, from seeing the additional complexity for 32bit systems in deciding the result dtype, I do wonder if we actually want to get rid of that system-dependent behaviour? (we also don't follow numpy's behaviour in the constructors, but always default to int64) 
Although given that we still call numpy for the actual operation, it probably doesn't reduce complexity to ensure this dtype guarantee (we would need to move the 32bit check to the implementation of the masked sum)